File: performance/ring/src/main/kotlin/org/jetbrains/ring/CallsBenchmark.kt
/*
 * Copyright 2010-2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.ring

private const val RUNS = 1_000_000

open class CallsBenchmark {

    interface I {
        fun foo(): Int

        fun foo1(): Int = 1
        fun foo2(): Int = 2
        fun foo3(): Int = 3
        fun foo4(): Int = 4
        fun foo5(): Int = 5
        fun foo6(): Int = 6
        fun foo7(): Int = 7
        fun foo8(): Int = 8
        fun foo9(): Int = 9
        fun foo10(): Int = 10
        fun foo11(): Int = 11
        fun foo12(): Int = 12
        fun foo13(): Int = 13
        fun foo14(): Int = 14
        fun foo15(): Int = 15
        fun foo16(): Int = 16
        fun foo17(): Int = 17
        fun foo18(): Int = 18
        fun foo19(): Int = 19
        fun foo20(): Int = 20
        fun foo21(): Int = 21
        fun foo22(): Int = 22
        fun foo23(): Int = 23
        fun foo24(): Int = 24
        fun foo25(): Int = 25
        fun foo26(): Int = 26
        fun foo27(): Int = 27
        fun foo28(): Int = 28
        fun foo29(): Int = 29
        fun foo30(): Int = 30
        fun foo31(): Int = 31
        fun foo32(): Int = 32
        fun foo33(): Int = 33
        fun foo34(): Int = 34
        fun foo35(): Int = 35
        fun foo36(): Int = 36
        fun foo37(): Int = 37
        fun foo38(): Int = 38
        fun foo39(): Int = 39
        fun foo40(): Int = 40
        fun foo41(): Int = 41
        fun foo42(): Int = 42
        fun foo43(): Int = 43
        fun foo44(): Int = 44
        fun foo45(): Int = 45
        fun foo46(): Int = 46
        fun foo47(): Int = 47
        fun foo48(): Int = 48
        fun foo49(): Int = 49
        fun foo50(): Int = 50
        fun foo51(): Int = 51
        fun foo52(): Int = 52
        fun foo53(): Int = 53
        fun foo54(): Int = 54
        fun foo55(): Int = 55
        fun foo56(): Int = 56
        fun foo57(): Int = 57
        fun foo58(): Int = 58
        fun foo59(): Int = 59
        fun foo60(): Int = 60
        fun foo61(): Int = 61
        fun foo62(): Int = 62
        fun foo63(): Int = 63
        fun foo64(): Int = 64
        fun foo65(): Int = 65
        fun foo66(): Int = 66
        fun foo67(): Int = 67
        fun foo68(): Int = 68
        fun foo69(): Int = 69
        fun foo70(): Int = 70
        fun foo71(): Int = 71
        fun foo72(): Int = 72
        fun foo73(): Int = 73
        fun foo74(): Int = 74
        fun foo75(): Int = 75
        fun foo76(): Int = 76
        fun foo77(): Int = 77
        fun foo78(): Int = 78
        fun foo79(): Int = 79
        fun foo80(): Int = 80
        fun foo81(): Int = 81
        fun foo82(): Int = 82
        fun foo83(): Int = 83
        fun foo84(): Int = 84
        fun foo85(): Int = 85
        fun foo86(): Int = 86
        fun foo87(): Int = 87
        fun foo88(): Int = 88
        fun foo89(): Int = 89
        fun foo90(): Int = 90
        fun foo91(): Int = 91
        fun foo92(): Int = 92
        fun foo93(): Int = 93
        fun foo94(): Int = 94
        fun foo95(): Int = 95
        fun foo96(): Int = 96
        fun foo97(): Int = 97
        fun foo98(): Int = 98
        fun foo99(): Int = 99
    }

    abstract class A : I

    open class B : A() {
        override fun foo() = 42
    }

    open class C : A() {
        override fun foo() = 117
    }

    class D: A() {
        override fun foo() = 314
    }

    open class X : A() {
        override fun foo() = 456456
    }

    open class Y : A() {
        override fun foo() = -398473
    }

    open class Z : A() {
        override fun foo() = 78298734
    }

    val d = D()
    val a1: A = B()
    val a2: A = C()
    val a3: A = d
    val i1: I = a1
    val i2: I = a2
    val i3: I = d
    val i4: I = X()
    val i5: I = Y()
    val i6: I = Z()

    fun finalMethodCall(): Int {
        var x = 0
        // TODO: optimize fields accesses
        val d = d
        for (i in 0 until RUNS)
            x += d.foo()
        return x
    }

    fun classOpenMethodCall_MonomorphicCallsite(): Int {
        var x = 0
        // TODO: optimize fields accesses
        val a1 = a1
        for (i in 0 until RUNS)
            x += a1.foo()
        return x
    }

    fun classOpenMethodCall_BimorphicCallsite(): Int {
        var x = 0
        // TODO: optimize fields accesses
        val a1 = a1
        val a2 = a2
        for (i in 0 until RUNS)
            x += (if (i and 1 == 0) a1 else a2).foo()
        return x
    }

    fun classOpenMethodCall_TrimorphicCallsite(): Int {
        var x = 0
        // TODO: optimize fields accesses
        val a1 = a1
        val a2 = a2
        val a3 = a3
        for (i in 0 until RUNS)
            x += (when (i % 3) {
                1 -> a1
                2 -> a2
                else -> a3
            }).foo()
        return x
    }

    fun interfaceMethodCall_MonomorphicCallsite(): Int {
        var x = 0
        // TODO: optimize fields accesses
        val i1 = i1
        for (i in 0 until RUNS)
            x += i1.foo()
        return x
    }

    fun interfaceMethodCall_BimorphicCallsite(): Int {
        var x = 0
        // TODO: optimize fields accesses
        val i1 = i1
        val i2 = i2
        for (i in 0 until RUNS)
            x += (if (i and 1 == 0) i1 else i2).foo()
        return x
    }

    fun interfaceMethodCall_TrimorphicCallsite(): Int {
        var x = 0
        // TODO: optimize fields accesses
        val i1 = i1
        val i2 = i2
        val i3 = i3
        for (i in 0 until RUNS)
            x += (when (i % 3) {
                1 -> i1
                2 -> i2
                else -> i3
            }).foo()
        return x
    }

    fun interfaceMethodCall_HexamorphicCallsite(): Int {
        var x = 0
        // TODO: optimize fields accesses
        val i1 = i1
        val i2 = i2
        val i3 = i3
        val i4 = i4
        val i5 = i5
        val i6 = i6
        for (i in 0 until RUNS)
            x += (when (i % 6) {
                1 -> i1
                2 -> i2
                3 -> i3
                4 -> i4
                5 -> i5
                else -> i6
            }).foo()
        return x
    }

    abstract class E {
        abstract fun foo(): Any
    }

    open class F : E() {
        override fun foo(): Int = 42
    }

    val e: E = F()

    fun returnBoxUnboxFolding(): Int {
        var x = 0
        // TODO: optimize fields accesses
        val e = e
        for (i in 0 until RUNS)
            x += e.foo() as Int
        return x
    }

    abstract class G<in T> {
        abstract fun foo(x: T): Int
    }

    open class H : G<Int>() {
        override fun foo(x: Int): Int {
            return x
        }
    }

    val g: G<Any> = H() as G<Any>

    fun parameterBoxUnboxFolding(): Int {
        var x = 0
        // TODO: optimize fields accesses
        val g = g
        for (i in 0 until RUNS)
            x += g.foo(i)
        return x
    }
}

File: performance/KotlinVsSwift/ring/src/CallsBenchmark.swift
/*
 * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
 * that can be found in the LICENSE file.
 */

import Foundation

protocol I {
    func foo() -> Int
}

protocol E {
    func foo() -> Any
}

extension I {
    func foo1() -> Int {
        return 1
    }
    func foo2() -> Int {
        return 2
    }
    func foo3() -> Int {
        return 3
    }
    func foo4() -> Int {
        return 4
    }
    func foo5() -> Int {
        return 5
    }
    func foo6() -> Int {
        return 6
    }
    func foo7() -> Int {
        return 7
    }
    func foo8() -> Int {
        return 8
    }
    func foo9() -> Int {
        return 9
    }
    func foo10() -> Int {
        return 10
    }
    func foo11() -> Int {
        return 11
    }
    func foo12() -> Int {
        return 12
    }
    func foo13() -> Int {
        return 13
    }
    func foo14() -> Int {
        return 14
    }
    func foo15() -> Int {
        return 15
    }
    func foo16() -> Int {
        return 16
    }
    func foo17() -> Int {
        return 17
    }
    func foo18() -> Int {
        return 18
    }
    func foo19() -> Int {
        return 19
    }
    func foo20() -> Int {
        return 20
    }
    func foo21() -> Int {
        return 21
    }
    func foo22() -> Int {
        return 22
    }
    func foo23() -> Int {
        return 23
    }
    func foo24() -> Int {
        return 24
    }
    func foo25() -> Int {
        return 25
    }
    func foo26() -> Int {
        return 26
    }
    func foo27() -> Int {
        return 27
    }
    func foo28() -> Int {
        return 28
    }
    func foo29() -> Int {
        return 29
    }
    func foo30() -> Int {
        return 30
    }
    func foo31() -> Int {
        return 31
    }
    func foo32() -> Int {
        return 32
    }
    func foo33() -> Int {
        return 33
    }
    func foo34() -> Int {
        return 34
    }
    func foo35() -> Int {
        return 35
    }
    func foo36() -> Int {
        return 36
    }
    func foo37() -> Int {
        return 37
    }
    func foo38() -> Int {
        return 38
    }
    func foo39() -> Int {
        return 39
    }
    func foo40() -> Int {
        return 40
    }
    func foo41() -> Int {
        return 41
    }
    func foo42() -> Int {
        return 42
    }
    func foo43() -> Int {
        return 43
    }
    func foo44() -> Int {
        return 44
    }
    func foo45() -> Int {
        return 45
    }
    func foo46() -> Int {
        return 46
    }
    func foo47() -> Int {
        return 47
    }
    func foo48() -> Int {
        return 48
    }
    func foo49() -> Int {
        return 49
    }
    func foo50() -> Int {
        return 50
    }
    func foo51() -> Int {
        return 51
    }
    func foo52() -> Int {
        return 52
    }
    func foo53() -> Int {
        return 53
    }
    func foo54() -> Int {
        return 54
    }
    func foo55() -> Int {
        return 55
    }
    func foo56() -> Int {
        return 56
    }
    func foo57() -> Int {
        return 57
    }
    func foo58() -> Int {
        return 58
    }
    func foo59() -> Int {
        return 59
    }
    func foo60() -> Int {
        return 60
    }
    func foo61() -> Int {
        return 31
    }
    func foo62() -> Int {
        return 62
    }
    func foo63() -> Int {
        return 63
    }
    func foo64() -> Int {
        return 64
    }
    func foo65() -> Int {
        return 65
    }
    func foo66() -> Int {
        return 66
    }
    func foo67() -> Int {
        return 67
    }
    func foo68() -> Int {
        return 68
    }
    func foo69() -> Int {
        return 69
    }
    func foo70() -> Int {
        return 70
    }
    func foo71() -> Int {
        return 71
    }
    func foo72() -> Int {
        return 72
    }
    func foo73() -> Int {
        return 73
    }
    func foo74() -> Int {
        return 74
    }
    func foo75() -> Int {
        return 75
    }
    func foo76() -> Int {
        return 76
    }
    func foo77() -> Int {
        return 77
    }
    func foo78() -> Int {
        return 78
    }
    func foo79() -> Int {
        return 79
    }
    func foo80() -> Int {
        return 80
    }
    func foo81() -> Int {
        return 81
    }
    func foo82() -> Int {
        return 82
    }
    func foo83() -> Int {
        return 83
    }
    func foo84() -> Int {
        return 84
    }
    func foo85() -> Int {
        return 85
    }
    func foo86() -> Int {
        return 86
    }
    func foo87() -> Int {
        return 87
    }
    func foo88() -> Int {
        return 88
    }
    func foo89() -> Int {
        return 89
    }
    func foo90() -> Int {
        return 90
    }
    func foo91() -> Int {
        return 91
    }
    func foo92() -> Int {
        return 92
    }
    func foo93() -> Int {
        return 93
    }
    func foo94() -> Int {
        return 94
    }
    func foo95() -> Int {
        return 95
    }
    func foo96() -> Int {
        return 96
    }
    func foo97() -> Int {
        return 97
    }
    func foo98() -> Int {
        return 98
    }
    func foo99() -> Int {
        return 99
    }

}

protocol A : I {}

class CallsBenchmarks {
    class B : A {
        func foo() -> Int {
            return 42
        }
    }
    class C : A {
        func foo() -> Int {
            return 117
        }
    }
    class D : A {
        func foo() -> Int {
            return 314
        }
    }
    class X : A {
        func foo() -> Int {
            return 456456
        }
    }
    class Y : A {
        func foo() -> Int {
            return -398473
        }
    }
    class Z : A {
        func foo() -> Int {
            return 8298734
        }
    }
    
    let d = D()
    let a1: A = B()
    let a2: A = C()
    lazy var a3: A = d
    lazy var i1: I = a1
    lazy var i2: I = a2
    lazy var i3: I = d
    let i4: I = X()
    let i5: I = Y()
    let i6: I = Z()
    
    func finalMethodCall() -> Int {
        var x = 0
        let d = self.d
        for _ in 0..<Constants.RUNS {
            x += d.foo()
        }
        return x
    }
    
    func classOpenMethodCall_MonomorphicCallsite() -> Int {
        var x = 0
        let a1 = self.a1
        for _ in 0..<Constants.RUNS {
            x += a1.foo()
        }
        return x
    }
    
    func classOpenMethodCall_BimorphicCallsite() -> Int {
        var x = 0
        let a1 = self.a1
        let a2 = self.a2
        for i in 0..<Constants.RUNS {
            x += ((i & 1 == 0) ? a1 : a2).foo()
        }
        return x
    }
    
    func classOpenMethodCall_TrimorphicCallsite() -> Int {
        var x = 0
        
        let a1 = self.a1
        let a2 = self.a2
        let a3 = self.a3
        for i in 0..<Constants.RUNS {
            switch (i % 3) {
                case 1: x += a1.foo()
                case 2: x += a2.foo()
                default: x += a3.foo()
            }
        }
        return x
    }
    
    func interfaceMethodCall_MonomorphicCallsite() -> Int {
        var x = 0
        let i1 = self.i1
        for _ in 0..<Constants.RUNS {
            x += i1.foo()
        }
        return x
    }
    
    func interfaceMethodCall_BimorphicCallsite() -> Int {
        var x = 0
        let i1 = self.i1
        let i2 = self.i2
        for i in 0..<Constants.RUNS {
            x += ((i & 1 == 0) ? i1 : i2).foo()
        }
        return x
    }
    
    func interfaceMethodCall_TrimorphicCallsite() -> Int {
        var x = 0
        let i1 = self.i1
        let i2 = self.i2
        let i3 = self.i3
        for i in 0..<Constants.RUNS {
            switch (i % 3) {
                case 1: x += i1.foo()
                case 2: x += i2.foo()
                default: x += i3.foo()
            }
        }
        return x
    }
    
    func interfaceMethodCall_HexamorphicCallsite() -> Int {
        var x = 0
        let i1 = self.i1
        let i2 = self.i2
        let i3 = self.i3
        let i4 = self.i4
        let i5 = self.i5
        let i6 = self.i6
        for i in 0..<Constants.RUNS {
            switch (i % 6) {
                case 1: x += i1.foo()
                case 2: x += i2.foo()
                case 3: x += i3.foo()
                case 4: x += i4.foo()
                case 5: x += i5.foo()
                default: x += i6.foo()
            }
        }
        return x
    }
        
    class F : E {
        func foo() -> Any {
            return 42
        }
    }
    
    let e: E = F()
    
    func returnBoxUnboxFolding() -> Int {
        var x = 0
        let e = self.e
        for _ in 0..<Constants.RUNS {
            x += e.foo() as! Int
        }
        return x
    }
}
